### ISSUE TITLE
Type inference fails with SAM and wildcard subtyping, works in 2.13

### ISSUE THREAD

**Issue by bishabosha** [lampepfl/dotty#18096]
see [original forum post](https://contributors.scala-lang.org/t/better-type-inference-for-scala-send-us-your-problematic-cases/2410/183) reported by @mdedetrich

## Compiler version

3.3.0

## Minimized code

```Scala
package pekko

// package org.apache.pekko.japi.function
@FunctionalInterface
trait Function[-T, +R] extends java.io.Serializable {
  @throws(classOf[Exception])
  def apply(param: T): R
}

// package org.apache.pekko.javadsl
final class Source[Out, Mat]() {
  def mapConcat[T](f: pekko.Function[Out, _ <: java.lang.Iterable[T]]): Source[T, Mat] =
    ???
}

// package com.google.pubsub.v1
final class ReceivedMessage

// package com.google.pubsub.v1
final class StreamingPullResponse {
  def getReceivedMessagesList: java.util.List[ReceivedMessage] = ???
}

// package example
class Materialised

// package example
object Test {
  val streamingPullResult: Source[StreamingPullResponse, Materialised] = ???

  def test =
    streamingPullResult
      .mapConcat(((response: StreamingPullResponse) => response.getReceivedMessagesList))
}
```

## Output

```scala
Compiling project (Scala 3.3.0, JVM)
[error] ./foo.scala:37:21
[error] Found:    pekko.StreamingPullResponse => java.util.List[pekko.ReceivedMessage]
[error] Required: pekko.Function[pekko.StreamingPullResponse, ? <: Iterable[T]]
[error] 
[error] where:    T is a type variable
[error]         .mapConcat(((response: StreamingPullResponse) => response.getReceivedMessagesList))
[error]                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

adding the explicit type `pekko.Function[StreamingPullResponse, java.util.List[pekko.ReceivedMessage]]` fixes the error,

also inference works if the signature of mapConcat changes to
```scala
def mapConcat[F[X] <: java.lang.Iterable[X], T](f: pekko.Function[Out, F[T]]): Source[T, Mat]
```

## Expectation

it should compile, like it did in scala 2.13.10

**Comment by bishabosha** [comment]
@smarter asking your opinion here

**Comment by mdedetrich** [comment]
> compiling with scala 2.13.10 works

Shouldn't this be `compiling with scala 3 works`?

**Comment by He-Pin** [comment]
@mdedetrich or `as compiling with scala 2.13.10 works`

**Comment by bishabosha** [comment]
I updated the expectation, sorry for the confusion

**Comment by dwijnand** [comment]
Minimised further

```scala
trait F1[-T1, +R] extends AnyRef { def apply(v1: T1): R }
class R { def l: List[Any] = ??? }
class S { def m[T](f: F1[R, ? <: List[T]]): S = ??? }
class T1 { def t1(s: S) = s.m((r: R) => r.l) }
```

**Comment by smarter** [comment]
I'm working on a fix for #16065 that also fixes this issue.
